# Cross Validation Techniques

In [1]:
import pandas as pd

<h4> About Dataset </h4>

This is a cancer dataset, there are different features based on that need to predict what kind of cancer it is!

In [3]:
df = pd.read_csv("E:\M60\ml_dataset\cancer_dataset.csv")
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [4]:
# Indepdent & dependent features
X = df.iloc[:, 2:]       # indepdenden (from col 2 to all)
y = df.iloc[:,1]         # dependent (diagnosis)

In [5]:
X.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   fractal_dimension_mean  ...  texture_worst  perimeter_worst  area_worst  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  Unnamed: 32  
0          0.4601                  0.11890          NaN  
1          0.2750                  0.08902          NaN  
2          0.3613                  0.08758          NaN  
3          0.6638                  0.17300          NaN  
4          0.2364                  0.07678          NaN  

[5 rows x 31 columns]

- last col (unnamed32) which basically null, so we can drop it by dropna

In [6]:
X = X.dropna(axis=1)

In [7]:
X.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   fractal_dimension_mean  ...  radius_worst  texture_worst  perimeter_worst  \
0                 0.07871  ...         25.38          17.33           184.60   
1                 0.05667  ...         24.99          23.41           158.80   
2                 0.05999  ...         23.57          25.53           152.50   
3                 0.09744  ...         14.91          26.50            98.87   
4                 0.05883  ...         22.54          16.67           152.20   

   area_worst  smoothness_worst  compactness_worst  concavity_worst  \
0      2019.0            0.1622             0.6656           0.7119   
1      1956.0            0.1238             0.1866           0.2416   
2      1709.0            0.1444             0.4245           0.4504   
3       567.7            0.2098             0.8663           0.6869   
4      1575.0            0.1374             0.2050           0.4000   

   concave points_worst  symmetry_worst  fractal_dimension_worst  
0                0.2654          0.4601                  0.11890  
1                0.1860          0.2750                  0.08902  
2                0.2430          0.3613                  0.08758  
3                0.2575          0.6638                  0.17300  
4                0.1625          0.2364                  0.07678  

[5 rows x 30 columns]

In [8]:
y.head()

0    M
1    M
2    M
3    M
4    M
Name: diagnosis, dtype: object

In [11]:
# Check whether the dataset is balance or imbalanced, lets check with target col (B:benign, M:malignant), its pretty imbalanced
y.value_counts()

diagnosis
B    357
M    212
Name: count, dtype: int64

<center> <h3> 1.0 HoldOut Validation Technique - Train Test & Split </h3> </center>

- Hold out validation techniue its kind of traditional train test split we do regularly
- We train our model with train dataset & validate our model with the help of test dataset

#### 1.1 Excercise

In [15]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=0)
model = DecisionTreeClassifier()
model.fit(X_train, y_train)
result = model.score(X_test, y_test)
print(result)

0.9181286549707602


<center> <h3> 2.0 KFold Validation Technique </h3></center>

#### 2.1. Expercise

In [16]:
from sklearn.model_selection import KFold
model = DecisionTreeClassifier()
kfold_validation = KFold(10)

In [17]:
import numpy as np
from sklearn.model_selection import cross_val_score
results = cross_val_score(model, X,y, cv=kfold_validation)
print(results)           # return 10 accuracy in a list/ numpy array
print(np.mean(results))  # mean of result

[0.9122807  0.9122807  0.87719298 0.94736842 0.9122807  0.96491228
 0.89473684 0.98245614 0.98245614 0.94642857]
0.9332393483709274


<center> <h3> 3.0 Stratified K-fold Cross Validation Technique </h3> </center>

In [20]:
from sklearn.model_selection import StratifiedKFold
skfold = StratifiedKFold(n_splits=5) # as per data size is n_split=5 is pretty good
model = DecisionTreeClassifier()
scores = cross_val_score(model, X,y, cv=skfold)
print(scores)
print(np.mean(scores))

[0.90350877 0.90350877 0.92982456 0.92105263 0.89380531]
0.9103400093153237


<center> <h3> 4.0 Leave One Out Cross Validation (LOOCV) </h3> </center>

<center><h2> Code Narration (Expercise 1.0) </h2></center>

1. Import decision Tree
2. import train test split
3. train test split parameterized, it will take (indepdent, dependent, test_size, random_state) 4 params
4. Inialize the model / decision tree classfier with default param, you can provide in hyperparameter at cross val
5. fit/train the model with train data ( X, y)
6. model score/ result
7. print the result

### Disadvance / limitation of train test split

1. Here sinlge type of train test split that is defined by randome state, its select/ split (train/test) randomly, see if you change the radom state value result will be vary
2. We dont know how the splitting happend randomly and get diffent accuracy, it may leads to overfit & underfit as also
3. To overcome this we can go to K fold cross validation

#### K fold Cross Validation

1. Here we can specify the folds, say told 10 folds, it will split into 10 folds (train & test in each folds) and apply ml model for all this splits, then will try to see what the accuracy we are getting
2. Finally, we will take the average mean of all the accuracy and confirmed to stakeholders
3. if Kfold=10 that tell us it will split the dataset into 10 folds and get 10 diffenet accuracy and finally take mean of this 10 accuracy; while in excercise it split randomly (train & test with no folds concept and get a single accuracy)
4. this diffent accuracy (i.g 10) need to verify and pass into model and see the accuracy it does by 'cross_val_score' from sklearn model selection
5. in cross_val_score we will pass (model, independent & dependent variable, cv=kfold_validation); here cv refer how many split are there, so the result will  be in a list of 10 different accuracy
6. finally np.mean of that 10 different accuracy

#### Result Interpretation ( K Fold Validation)

1. accuracy range (.87 - .98) out of 10 accuracy
2. Say this accuracy to stakeholders unless or untill this dataset is not balanced
3. here least is 87 & max is 98 and mean is 93, tell stackholder this is pretty good, untill or unless dataset is imbalanced!
4. if the range is huge then again you have to do hypterparameter tunning & if the range is small no need to do hyperparameter tuning

#### Stratified K-fold Cross Validation

1. It is useful then the dataset is imbalanced
2. For example, in a binary classification one category is 900 and second category is 100  (highly imbalanced) so, this ratio is nothing but 9:1, in that case if you go for train test split (i.g Kfold) then correct proportion will not be splited or picked up for spliting in respect to both the category it may be the situation that all the data taken from a particular category
3. So, Stratified K-fold validation make sure that, if the dataset is imblanced, it will try to take equal proportion of data from both the category, for example for a case Yes:700, No:300, it will try to take around 400 hundred from one category and 300 from another categtory (for example test_size =.70)
4. try to take proportionately as best as possible (Approximately close), so Stratifed K-fold cross validation will be diffinitely used in <b> Imbalanced dataset</b>

#### @ Credits: Krish Naik